File: sort/sort.cpp
#include <iostream>
#define SWAP(A,B,temp) ( temp = A, A=B, B=temp )
using namespace std;

void selectiocSort(int* list, int n){

    int minIndex, temp;

    for( int i = 0; i < n-1 ; i++){
        minIndex = i;

        for(int j = i+1 ; j < n ; j++){
            if(list[j] < list[minIndex]){
                minIndex = j;
            }
        }

        if( minIndex != i )
            SWAP(list[minIndex], list[i], temp);
    }
}

void bubbleSort(int* list, int n){

    int temp;

    for( int i = 0 ; i < n ; i++){
        for( int j = 0 ; j < n - i -1 ; j++){
            if( list[j] > list[j+1])
                SWAP(list[j], list[j+1], temp);
        }
    }
}

void insertionSort(int* list, int n){
    int temp;

    for(int i = 1 ; i < n ; i++){
        
        int key = list[i];

        for(int j = i -1 ; (j >= 0 && key < list[j] ) ; j-- ){
            list[j+1] = list[j];
        }
        list[i] = key;   
    }
}

// pivot값 중심으로 왼-작, 오-큰 부분 배열로 분할, pivot의 위치 반환
int partition(int* list, int left, int right){

    int low, high, temp;
    int pivot = list[left];

    low = left+1;
    high = right;

    //왼-작, 오-큰 으로 분할해야하니까 왼-큰, 오-작 을 찾아야함
    while(low <= high){
        while( list[low] < pivot && low <= right ) low++;
        while( list[high] > pivot && high <= left ) right--;

        // 왼-큰, 오-작 인 경우 바꿔주기
        if(low <= high){
            SWAP(list[low], list[high], temp);

            //바꿔줬으므로 인덱스는 이동하게끔
            low++;
            right--;
        }
    }
    SWAP(list[left], list[high], temp);

    return high;
}

void quickSort(int* list, int left, int right){
    int pivotIndex;

    if(left < right){
            pivotIndex = partition(list, left, right);
    }
    quickSort(list, left, pivotIndex -1);
    quickSort(list, pivotIndex+1, right);

}

int* merge(int* list, int left, int mid, int right){
    int* sorted = (int*)malloc( (*list) );

    int low = left;
    int high = mid +1;
    int index = left;

    while(low <= mid && high <= right ){
        if(list[mid] < list[high])
            sorted[index++] = list[low++];
        else
            sorted[index++] = list[high++];
    }

    // 남아있는 데이터가 있는 경우 ( 즉, 어느 한 쪽이 다 들어간 경우 )
    if( low > mid){
        for( int i = high ; i <= right ; i++)
            sorted[index++] = list[i];
    }
    else{
        for( int i = low ; i <= mid ; i++)
            sorted[index++] = list[i];
    }

    return sorted;
}


void mergeSort(int* list, int left, int right){

    int* sorted;

    if( left < right){
        int mid = ( left + right ) / 2; // Divide
        mergeSort(list, left, mid -1); // Conquer
        mergeSort(list, mid, right); // Conquer
        sorted = merge(list, left, mid, right);
    }    

}

void heapify(int* list, int index, int size){

    int temp;
    int largest = index;
    
    int leftChild = (index*2);
    int rightChild = (index*2+1);

    if( leftChild < size && list[leftChild] > list[largest])
        largest = leftChild;
    if( rightChild < size && list[rightChild] > list[largest])
        largest = rightChild;
    
    // 위치가 바뀌어야 한다면
    if(largest != index){
        SWAP(list[index], list[largest], temp);

        // 자리가 바뀐 자식에 의해 한 번 더 heapify
        heapify(list, largest, size);
    }
}

void heapSort(int* list, int size){
    for(int i = size/2 ; i >= 1 ; i--){
        heapify(list, i, size);
    }
}

int main(void){

    int ary[6] = {0, 1, 2, 3, 4, 5};

    heapSort(ary, 6);
    
    for(int i = 1 ; i < 6 ; i++)
        cout << ary[i];

    return 0;
}

File: sort/sort.py
def bubble(ary):
    for i in range(len(ary)-1):
        for j in range(len(ary)-1-i):
            if(ary[j] > ary[j+1]):
                temp = ary[j]
                ary[j] = ary[j+1]
                ary[j+1] = temp
    
    print(ary)


# 좀 더 쉽게 생각해보자잔자리장
def insertion(ary):
    

    for i in range(1, len(ary)):
        # 2번쨰 요소부터 시작
        key = ary[i]
        idx = 0

        # 끝에서부터 돌면서 삽입 위치 찾기
        for j in range(i, -1, -1):
            if(key > ary[j]):
                idx = j
            else:
                ary[j] = ary[j-1]

        
        ary[idx] = key

    print(ary)
    

def selection(ary):

    for i in range(len(ary)):
        idx = i # 해당 위치에 들어간다

        for j in range(i+1, len(ary)):
            if(ary[idx] > ary[j]):
                idx = j

        temp = ary[i]
        ary[i] = ary[idx]
        ary[idx] = temp

    print(ary)





    

def selection(ary):

    for i in range(len(ary)):
        
        # i번쨰 요소와 바꾼다.
        # 이후의 배열 들에서 최소값 인덱스 찾기
        idx = i
        for j in range(i+1, len(ary)):
            if(ary[idx] > ary[j]):
                idx = j
        
        temp = ary[i]
        ary[i] = ary[idx]
        ary[idx] = temp
    
    print(ary)




def quick(ary, left, right):

    if(left < right):
        idx = partition(ary, left, right)
        quick(ary, left, idx)
        quick(ary, idx+1, right)

    return ary

def partition(ary, left, right):
    i = left
    pivot = ary[left]
    #print("pivot", pivot, ary)
    for j in range(left+1, right):
        if(ary[j]<pivot):
            i +=1
            temp = ary[j]
            ary[j] = ary[i]
            ary[i] = temp
            #print("part",ary)

    temp = ary[i]
    ary[i] = pivot
    ary[left] = temp

    return i






sorted=[]
def merge(ary, left, right):

    if(right-left < 2):
        return
    mid = (left+right) // 2
   
    merge(ary, left, mid)
    merge(ary, mid, right)

    # 2개의 부분 배열 합병
    _merge(ary, left, mid, right)

def _merge(ary, left, mid, right):
    temp = []
    l, h = left, right
    print("enter", left, mid, right)
    while(l < mid and h < right):
        if (ary[l]< ary[h]):
            temp.append(ary[l])
            l +=1
        else:
            temp.append(ary[h])
    
        print("whie", temp)
        while l < mid:
            temp.append(arr[l])
            l += 1
        while h < high:
            temp.append(arr[h])
            h +=1

    print(temp)
ary = [6, 10, 13, 5, 8, 3, 2, 11]
#bubble(ary)
#insertion2(ary)
#insertion(ary)
#selection(ary)


end = quick(ary, 0, len(ary))
print(end)

#merge(ary, 0, len(ary))
#print(sorted)

#print((2+2)/2)
